go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -2438,6 +2440,58 @@ func (s *VtctldServer) GetUnresolvedTransactions(ctx context.Context, req *vtctl
 	}, nil
 }
 
+// ConcludeTransaction is part of the vtctlservicepb.VtctldServer interface.
+// It concludes the unresolved distributed transaction.
+func (s *VtctldServer) ConcludeTransaction(ctx context.Context, req *vtctldatapb.ConcludeTransactionRequest) (resp *vtctldatapb.ConcludeTransactionResponse, err error) {
+	span, ctx := trace.NewSpan(ctx, "VtctldServer.ConcludeTransaction")
+	defer span.Finish()
+
+	span.Annotate("dtid", req.Dtid)
+	span.Annotate("participants", req.Participants)
+
+	ss, err := dtids.ShardSession(req.Dtid)
+	if err != nil {
+		return nil, err
+	}
+	primary, err := s.getPrimaryTablet(ctx, ss.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	eg, newCtx := errgroup.WithContext(ctx)
+	eg.SetLimit(10)
+	for _, rm := range req.Participants {
+		eg.Go(func() error {
+			primary, err := s.getPrimaryTablet(newCtx, rm)
+			if err != nil {
+				return err
+			}
+			return s.tmc.ConcludeTransaction(newCtx, primary.Tablet, req.Dtid, false)
+		})
+	}
```

Comment:
What if the `ss.Target` tablet is part of the requested participants. Would we not end up calling ConcludeTransaction on it twice? First with the last argument being false, and the next time with true.

Also, if the `ss.Target` is not part of the requested participants, should we even do the second `ConcludeTransaction` call?

Reply:
That is the Metadata Manager Shard, it's conclude is about deleting the Transaction record once everything is resolved in other participating shards.

Participating shards is exclusive of Metadata Manager shard.